src/main/java/io/cdap/plugin/gcp/gcs/sink/GCSBatchSink.java

```diff
@@ -82,18 +111,95 @@ public void prepareRun(BatchSinkContext context) throws Exception {
     if (bucket == null) {
       GCPUtils.createBucket(storage, config.getBucket(), config.getLocation(), cmekKey);
     }
+
   }
 
   @Override
   protected Map<String, String> getFileSystemProperties(BatchSinkContext context) {
     return GCPUtils.getFileSystemProperties(config, config.getPath(), new HashMap<>());
   }
 
+  @Override
+  protected String getOutputDir(long logicalStartTime) {
+    this.outputPath = super.getOutputDir(logicalStartTime);
+    return this.outputPath;
+  }
+
   @Override
   protected void recordLineage(LineageRecorder lineageRecorder, List<String> outputFields) {
     lineageRecorder.recordWrite("Write", "Wrote to Google Cloud Storage.", outputFields);
   }
 
+  @Override
+  public void onRunFinish(boolean succeeded, BatchSinkContext context) {
+    super.onRunFinish(succeeded, context);
+    recordMetric(succeeded, context);
+  }
+
+  private void recordMetric(boolean succeeded, BatchSinkContext context) {
+    if (!succeeded) {
+      return;
+    }
+
+    try {
+      recordMetricForAllFiles(context);
+    } catch (Exception e) {
+      LOG.warn("Metrics for the number of affected rows maybe incorrect.", e);
+    }
+  }
+
+  private void recordMetricForAllFiles(BatchSinkContext context) throws IOException {
+    GCSPath gcsPath = GCSPath.from(outputPath);
+    //add all the configurations necessary to create file system
+    Configuration configuration = new Configuration();
+    getFileSystemProperties(context).entrySet().stream()
+      .forEach(entry -> configuration.set(entry.getKey(), entry.getValue()));
+
+    try (FileSystem fileSystem = getFileSystem(outputPath, configuration)) {
+      if (!(fileSystem instanceof GoogleHadoopFileSystem)) {
+        return;
+      }
+
+      GoogleCloudStorageFileSystem gcsFs = ((GoogleHadoopFileSystem) fileSystem).getGcsFs();
+      try {
+        //For each file, read metadata to get count and sum them up
+        List<FileInfo> fileInfos = gcsFs.listAllFileInfoForPrefix(gcsPath.getUri());
+        fileInfos.stream()
+          .forEach(fileInfo -> recordMetric(context.getMetrics(), extractRecordCount(fileInfo.getAttributes())));
+      } finally {
+        gcsFs.close();
```

Comment:
Is it safe to close since it actually is the underlying GcsFs of the FileSystem?

Reply:
Not applicable after changing code to use Storage API.